torchtune/modules/early_exit_loss.py

```diff
@@ -0,0 +1,334 @@
+        if self.train_last_layer:
+            do_output_hidden_states[-1] = True
+        return do_output_hidden_states
+
+
+class RotationalEarlyExitCurriculum(EarlyExitCurriculum):
+    """
+    A rotational early exit curriculum, which rotates the layer enablement one step forward
+    at each step.
+    Args:
+        do_output_hidden_states (List[bool]): A list indicating whether each layer's hidden state
+            should be output to calculate their losses.
+        max_steps (int): The maximum number of steps in the curriculum.
+        train_last_layer (bool, optional): Whether to always calculate loss for the last layer. Defaults to True.
+        last_step (Optional[int]): The last step the curriculum stopped at in a previous run.
+            This is used when resuming training.
+        verbose (bool, optional): Whether to print verbose logs. Defaults to False.
+    """
+
+    def __init__(
+        self,
+        do_output_hidden_states: List[bool],
+        max_steps: int,
+        train_last_layer: bool = True,
+        last_step: Optional[int] = None,
+        verbose: bool = False,
+    ):
+        super().__init__(
+            do_output_hidden_states=do_output_hidden_states,
+            max_steps=max_steps,
+            train_last_layer=train_last_layer,
+            last_step=last_step,
+            verbose=verbose,
+        )
+        self._initial_do_output_hidden_states = np.copy(self._do_output_hidden_states)
+
+    def step(self):
+        """
+        Rotate the layer enablement one step forward.
+        This method updates the `do_output_hidden_states` attribute by rotating it one position to the right.
+        """
+        # Rotate layer enablement one step forward
+        self._do_output_hidden_states = np.roll(self._do_output_hidden_states, 1)
+
+        self._step += 1
+        if self.verbose:
+            log.info(
+                f"Updated self._do_output_hidden_states to {self._do_output_hidden_states}."
+            )
+
+
+class GradualEarlyExitCurriculum(EarlyExitCurriculum):
+    """
+    A gradual early exit curriculum, which gradually enables more layers (starting from the last layer) as training progresses.
+    Args:
+        do_output_hidden_states (List[bool]): A list indicating whether each layer's hidden state
+            should be output to calculate their losses.
+        max_steps (int): The maximum number of steps in the curriculum.
+        train_last_layer (bool, optional): Whether to always calculate loss for the last layer. Defaults to True.
+        last_step (Optional[int]): The last step the curriculum stopped at in a previous run.
+            This is used when resuming training.
+        fraction_scale (float, optional): A scaling factor to determine at which fraction
+            of steps, all the layers will be enabled. At `steps = max_steps * fraction_scale`, all the layers will be enabled.
+        verbose (bool, optional): Whether to print verbose logs. Defaults to False.
```

Comment:
nit but in our docstrings I don't think we typically put optional unless we actually mean `Optional` type (i.e. `Union[T, None]`)

```suggestion
        train_last_layer (bool): Whether to always calculate loss for the last layer. Defaults to True.
        last_step (Optional[int]): The last step the curriculum stopped at in a previous run.
            This is used when resuming training.
        fraction_scale (float): A scaling factor to determine at which fraction
            of steps, all the layers will be enabled. At `steps = max_steps * fraction_scale`, all the layers will be enabled. Defaults to 0.5.
        verbose (bool): Whether to print verbose logs. Defaults to False.
```